components/omega/src/ocn/Eos.h

```diff
@@ -0,0 +1,251 @@
+#ifndef OMEGA_EOS_H
+#define OMEGA_EOS_H
+//===-- ocn/Eos.h - Equation of State --------------------*- C++ -*-===//
+//
+/// \file
+/// \brief Contains functors for calculating specific volume
+///
+/// This header defines functors to be called by the time-stepping scheme
+/// to calculate the specific volume based on the choice of EOS
+//
+//===----------------------------------------------------------------------===//
+
+#include "AuxiliaryState.h"
+#include "Config.h"
+#include "EosConstants.h"
+#include "HorzMesh.h"
+#include "MachEnv.h"
+#include "OceanState.h"
+#include "TimeMgr.h"
+#include "Tracers.h"
+
+#include "OmegaKokkos.h"
+
+namespace OMEGA {
+
+enum class EosType {
+   Linear,       /// Linear equation of state
+   TEOS10Poly75t /// Roquet et al. 2015 75 term expansion
+};
+
+//
+/// TEOS10 75-term polynomial
+class TEOS10Poly75t {
+ public:
+   Array2DReal specVolPcoeffs;
+
+   /// constructor declaration
+   TEOS10Poly75t();
+
+   KOKKOS_FUNCTION void operator()(const Array2DReal &SpecVol, I4 ICell,
+                                   I4 KChunk,
+                                   const Array2DReal &ConservativeTemperature,
+                                   const Array2DReal &AbsoluteSalinity,
+                                   const Array2DReal &Pressure) const {
+
+      const I4 KStart = KChunk * VecLength;
+      for (int KVec = 0; KVec < VecLength; ++KVec) {
+         const I4 K = KStart + KVec;
+         calcPCoeffs(specVolPcoeffs, KVec, ConservativeTemperature(ICell, K),
+                     AbsoluteSalinity(ICell, K));
+         SpecVol(ICell, K) = calcRefProfile(Pressure(ICell, K)) +
+                             calcDelta(KVec, Pressure(ICell, K));
+      }
+   }
+
+   // To be added:
+   // Note that it assumes that we have called calcPCoeffs already
+   //
+   // KOKKOS_FUNCTION void calcDisplacedSpecVol(const Array2DReal
+   // &SpecVolDisplaced,
+   //                                 I4 ICell, I4 KChunk,
+   //                                 const Array2DReal &Pressure) const {
+
+   //    const I4 KStart = KChunk * VecLength;
+   //    // insert exception for the surface
+   //    for (int KVec = 0; KVec < VecLength; ++KVec) {
+   //       const I4 K        = KStart + 1 + KVec;
+   //       SpecVolDisplaced(ICell, K) = calcRefProfile(Pressure(ICell, K-1)) +
+   //       calcDelta(Pressure(ICell, K-1));
+   //    }
+   // }
```

Comment:
Is there a reason not to add this in this PR?

Reply:
I just hadn't done it and needed to double check whether we wanted one level up or one level down. Implementing now but I don't have a test for it yet... 🤔